File: contest/vol115/11582.cpp
#include<bits/stdc++.h>
using namespace std;

int big_mod( unsigned long long base, unsigned long long exp, int mod )
{
	int mul[ 64 ] = { base % mod }, ans = 1;
	for( int i = 1; i < 64; ++i )
		mul[ i ] = mul[ i - 1 ] * mul[ i - 1 ] % mod;
	for( int i = 0; i < 64; ++i )
		if( exp & 1ull << i )
			ans = ans * mul[ i ] % mod;
	return ans;
}

int main()
{
	int t;
	unsigned long long base, exp, mod;

	for( cin >> t; t-- && cin >> base >> exp >> mod; )
	{
		vector<int> cycle( 1, 0 );
		int size = 2;
		cycle.push_back( 1 );

		while( ( cycle[ size - 1 ] + cycle[ size - 2 ] ) % mod || cycle[ size - 1 ] != 1 )
			cycle.push_back( ( cycle[ size - 1 ] + cycle[ size - 2 ] ) % mod ), ++size;
		cout << ( mod == 1? 0 : cycle[ big_mod( base, exp, size ) ] ) << endl;
	}
}

File: vol115/11582.py
# TLE

def big_mod( base, exp, mod ):
	mul, ans = [ base % mod ], 1 if mod > 1 else 0
	for i in range( 1, 64 ):
		mul.append( mul[ -1 ] ** 2 % mod )
	for i in range( 64 ):
		if exp & 1 << i != 0:
			ans = ans * mul[ i ] % mod
	return ans

for t in range( int( input() ) ):
	base, exp, mod = map( int, input().split() )
	if mod == 1:
		print( 0 )
		continue
	cycle = [ 0, 1 ]
	while sum( cycle[ -2: ] ) % mod != 0 or cycle[ -1 ] != 1:
		cycle.append( sum( cycle[ -2: ] ) % mod )
	print( cycle[ big_mod( base, exp, len( cycle ) ) ] )
